privval/ipc.go

```diff
@@ -0,0 +1,218 @@
+package privval
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/types"
+)
+
+// IPCPV implements PrivValidator, it uses a unix socket to request signatures
+// from an external process.
+type IPCPV struct {
+	cmn.BaseService
+	*RemoteSignerClient
+
+	addr string
+
+	connTimeout time.Duration
+
+	lock       sync.Mutex
+	cancelPing chan bool
+}
+
+// Check that IPCPV implements PrivValidator.
+var _ types.PrivValidator = (*IPCPV)(nil)
+
+// NewIPCPV returns an instance of IPCPV.
+func NewIPCPV(
+	logger log.Logger,
+	socketAddr string,
+) *IPCPV {
+	sc := &IPCPV{
+		addr:        socketAddr,
+		connTimeout: connTimeout,
+	}
+
+	sc.BaseService = *cmn.NewBaseService(logger, "IPCPV", sc)
+	sc.RemoteSignerClient = NewRemoteSignerClient(logger, nil)
+
+	return sc
+}
+
+// OnStart implements cmn.Service.
+func (sc *IPCPV) OnStart() error {
+	err := sc.connect()
+	if err != nil {
+		err = cmn.ErrorWrap(err, "failed to connect")
+		sc.Logger.Error(
+			"OnStart",
+			"err", err,
+		)
+
+		return err
+	}
+
+	err = sc.RemoteSignerClient.Start()
```

Comment:
Nested lifecycle management seems undesirable, see general comment about the abstraction of the rs client.